### PR TITLE
[Converter] Added IsInRangeConverter

### PR DESCRIPTION
**Pull request by GUOLDEV**
I have created a simple converter to check if a value is between two values, returning true if the value is within the range and false if the value is out of the range. 

### Description of Change ###

The converter checks if the value is between minValue and maxValue, returning true if the value is within the range and false if the value is out of the range. 

### Bugs Fixed ###

- Fixes #

NA

### API Changes ###

<!-- List all API changes here (or just put None), example:

Added: 
 
- `string Class.Property { get; set; }`
- `void Class.Method();`

Changed:

 - `object Cell.OldPropertyName` => `object Cell.NewPropertyName`
 
If there is an entirely new API, then you can use a more verbose style:

```csharp
public static class NewClass {
    public static int SomeProperty { get; set; }
    public static void SomeMethod(string value);
}
``` -->

Added:

- `public object MinValue { get; set; }`
- `public object MaxValue { get; set; }`
- `public object Convert(object value, Type targetType, object parameter, CultureInfo culture)`

### Behavioral Changes ###

<!-- Describe any non-bug related behavioral changes that may change how users app behaves when upgrading to this version of the codebase. -->

![image](https://user-images.githubusercontent.com/22019245/115103340-93e27480-9f94-11eb-93a8-cdd540b57a47.png)

![image](https://user-images.githubusercontent.com/22019245/115103370-bffdf580-9f94-11eb-9011-4edca6665712.png)

### PR Checklist ###
<!-- Please check all the things you did here and double-check that you got it all, or state why you didn't do something -->
- [x] Has a linked Issue, and the Issue has been `approved`
- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard
<!-- If at all possible, please update/add the documentation on the repo below. We would very much appreciate that. If you are unable to, please consider at least opening an issue on the repo below so we know that Docs still need to be adjusted/created. Thanks! <3 -->
- [x] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

File: samples/XCT.Sample.Android/Xamarin.CommunityToolkit.Sample.Android.csproj [hunk 1]
```diff
@@ -31,6 +31,8 @@
     <DefineConstants>DEBUG;</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
+    <AndroidLinkMode>None</AndroidLinkMode>
+    <EmbedAssembliesIntoApk>true</EmbedAssembliesIntoApk>
```
Comment (AndreiMisiukevich) [inline samples/XCT.Sample.Android/Xamarin.CommunityToolkit.Sample.Android.csproj:35]: Please discard

File: samples/XCT.Sample.iOS/Xamarin.CommunityToolkit.Sample.iOS.csproj [hunk 2]
```diff
@@ -25,7 +25,7 @@
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
     <MtouchArch>x86_64</MtouchArch>
-    <MtouchLink>SdkOnly</MtouchLink>
+    <MtouchLink>None</MtouchLink>
```
Comment (AndreiMisiukevich) [inline samples/XCT.Sample.iOS/Xamarin.CommunityToolkit.Sample.iOS.csproj:28]: please discard

File: samples/XCT.Sample.iOS/Xamarin.CommunityToolkit.Sample.iOS.csproj [hunk 3]
```diff
@@ -61,6 +61,7 @@
     <MtouchArch>ARM64</MtouchArch>
     <CodesignKey>iPhone Developer</CodesignKey>
     <CodesignEntitlements>Entitlements.plist</CodesignEntitlements>
+    <MtouchLink>SdkOnly</MtouchLink>
```
Comment (AndreiMisiukevich) [inline samples/XCT.Sample.iOS/Xamarin.CommunityToolkit.Sample.iOS.csproj:64]: please discard

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/DateTimeInRangeConverter.shared.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Globalization;
+using Xamarin.CommunityToolkit.Extensions.Internals;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.Converters
+{
+	public class DateTimeInRangeConverter : ValueConverterExtension, IValueConverter
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/DateTimeInRangeConverter.shared.cs:8]: What is the purpose of this converter? why can't we use IsInRangeConverter?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Globalization;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.Converters
+{
+	public class IsInRangeConverter : BindableObject, IValueConverter
+	{
+		public bool RevertResult { get; set; } = false;
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs:9]: Personally, I am against of "RevertResult" property approach.
Let's use IsNotInRangeConverter approach

CC @pictos @jfversluis

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+		public static readonly BindableProperty MinValueProperty = BindableProperty.Create(nameof(MinValue), typeof(object), typeof(IsInRangeConverter), null);
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs:11]:
```suggestion
		public static readonly BindableProperty MinValueProperty = BindableProperty.Create(nameof(MinValue), typeof(object), typeof(IsInRangeConverter));
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+		public object MinValue
+		{
+			get => GetValue(MinValueProperty);
+			set => SetValue(MinValueProperty, value);
+		}
+
+		public static readonly BindableProperty MaxValueProperty = BindableProperty.Create(nameof(MaxValue), typeof(object), typeof(IsInRangeConverter), null);
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs:19]:
```suggestion
		public static readonly BindableProperty MaxValueProperty = BindableProperty.Create(nameof(MaxValue), typeof(object), typeof(IsInRangeConverter));
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+				throw new ArgumentException("Value cannot be null", nameof(MinValue));
+
+			if (MaxValue == null)
+				throw new ArgumentException("Value cannot be null", nameof(MaxValue));
+
+			if (value.GetType() != MinValue.GetType())
+				throw new ArgumentException("Values cannot be different types", nameof(MinValue));
+
+			if (value.GetType() != MaxValue.GetType())
+				throw new ArgumentException("Values cannot be different types", nameof(MinValue));
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs:42]: why I cannot compare **int** and **long** for example?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Globalization;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.Converters
+{
+	public class IsInRangeConverter : BindableObject, IValueConverter
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs:7]: I am not sure if it's a good idea to make Converters extending bindableObjects

CC @jfversluis ?

Comment (AndreiMisiukevich) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs:7]: If it's a good idea, then probably we should make ValueConverterExtension extending BindableObject. And this class should be inherited from ValueConverterExtension

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs (hunk 5 above)
Comment (pictos) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs:9]: Agreed

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs (hunk 9 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs:7]: @AndreiMisiukevich do you prefer static values?

File: samples/XCT.Sample.Android/Xamarin.CommunityToolkit.Sample.Android.csproj (hunk 1 above)
Comment (GUOLDEV) [reply samples/XCT.Sample.Android/Xamarin.CommunityToolkit.Sample.Android.csproj]: Done!

File: samples/XCT.Sample.iOS/Xamarin.CommunityToolkit.Sample.iOS.csproj (hunk 2 above)
Comment (GUOLDEV) [reply samples/XCT.Sample.iOS/Xamarin.CommunityToolkit.Sample.iOS.csproj]: Done

File: samples/XCT.Sample.iOS/Xamarin.CommunityToolkit.Sample.iOS.csproj (hunk 3 above)
Comment (GUOLDEV) [reply samples/XCT.Sample.iOS/Xamarin.CommunityToolkit.Sample.iOS.csproj]: Done

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/DateTimeInRangeConverter.shared.cs (hunk 4 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/DateTimeInRangeConverter.shared.cs]: It was just to exemplify the concept of this converter. I have removed it.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs (hunk 5 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs]: Done

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs (hunk 8 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs]: Done

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs (hunk 9 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs:7]: Hi @AndreiMisiukevich, who can help us to decide if we go with BindableObject or actually extend ValueConverterExtension?

Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs]: @pictos @AndreiMisiukevich I have made the changes based on the feedback, could you please let me know what you think?

Comment (pictos) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs]: @GUOLDEV sorry for the late reply, rush week. I'll re-review ASAP.  I did a quick look and looks good.

Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs]: Thanks @pictos for the reply, I am in the same situation, having a busy week, I am taking baby steps, but I will get there 😅👍

**Comment by GUOLDEV** [comment]
@AndreiMisiukevich @pictos It would be great if I can get an update about this converter, if we can go ahead as it is or if I need to make changes. Thanks

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Extensions/BindableObjectExtension.shared.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Extensions/BindableObjectExtension.shared.cs:1]: I don't think we need this class. Please drop it.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using System.Globalization;
+using Xamarin.CommunityToolkit.Extensions;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.Converters
+{
+	public class IsInRangeConverter : BindableObjectExtension, IValueConverter
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs:8]: It should be inherited from "ValueConverterExtension"

Just make ValueConverterExtension inherited from BindableObject

```csharp
public abstract class ValueConverterExtension : BindableObject, IMarkupExtension<IValueConverter>
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Extensions/BindableObjectExtension.shared.cs (hunk 10 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Extensions/BindableObjectExtension.shared.cs]: Done

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs (hunk 11 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs]: Done

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs (hunk 9 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs]: I have made ValueConverterExtension inherited from BindableObject.

**Comment by GUOLDEV** [comment]
Working on the documentation

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Extensions/ValueConverterExtension.shared.cs [hunk 12]
```diff
@@ -4,7 +4,7 @@
 
 namespace Xamarin.CommunityToolkit.Extensions.Internals
 {
-	public abstract class ValueConverterExtension : IMarkupExtension<IValueConverter>
+	public abstract class ValueConverterExtension : BindableObject, IMarkupExtension<IValueConverter>
```
[the remark this answers is not in the document]
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Extensions/ValueConverterExtension.shared.cs:7]: Hi @jfversluis it's to make properties in the Converter Bindable. In the case of IsInRangeConverter, MinValue and MaxValue can be changed programmatically binging it to a property in the ViewModel for example.

[the remark this answers is not in the document]
Comment (AndreiMisiukevich) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Extensions/ValueConverterExtension.shared.cs:7]: @jfversluis like this https://github.com/xamarin/XamarinCommunityToolkit/pull/1158/files#diff-9b74f20e840d4a2fa011040c31dd2aa295a9a183e416be422c7392a2bb49997bR17-R18

**Review by jfversluis — APPROVED** [review]

**Pull request merged**